### ISSUE TITLE
Investigate invalid use of ground truth which leads to improved performances

### ISSUE THREAD

**Issue by NathanMolinier** [spinalcordtoolbox/disc-labeling-hourglass#23]
## Description

Recently, I noticed that the post processing function applied during testing: [extract_skeleton](https://github.com/spinalcordtoolbox/disc-labeling-hourglass/blob/main/src/dlh/utils/test_utils.py#L28), used some [information](https://github.com/spinalcordtoolbox/disc-labeling-hourglass/blob/main/src/dlh/utils/test_utils.py#L37-L40) from the labels to know how many intervertebral discs were present in the image. Therefore, by doing this the performances of the hourglass are biased and less subject to false detection.

INVALID [step](https://github.com/spinalcordtoolbox/disc-labeling-hourglass/blob/main/src/dlh/utils/train_utils.py#L123-L129): Images with a discs number higher than the number of discs the hourglass was trained to find were also removed from the testing and the training.

## Conclusion

Further investigation need to be done in the function [extract_skeleton](https://github.com/spinalcordtoolbox/disc-labeling-hourglass/blob/main/src/dlh/utils/test_utils.py#L28) and in [create_skeleton](https://github.com/spinalcordtoolbox/disc-labeling-hourglass/blob/main/src/dlh/utils/skeleton.py#L14) to see how this post processing function could be improved/fixed to increase fairly the hourglass performances.

**Comment by NathanMolinier** [comment]
By removing this ground truth use, I noticed that some issues and false detections started to appear. Indeed, one problem is that some classes corresponding to unseen discs during training are detected which is causing errors in the skeleton reconstruction. Here is an example of a mask returned by the hourglass network corresponding to the disc 14 which is not present in the training set.
<img width="256" alt="Capture d’écran 2023-06-26 à 17 09 37" src="https://github.com/spinalcordtoolbox/disc-labeling-hourglass/assets/68945192/143725cf-4eb5-4852-a5fc-c518cfbe401f">

**Comment by NathanMolinier** [comment]
The best approach to avoid such false detections seems to be to reduce the number of class. Indeed, if we look at the spinegeneric dataset,

For T1w images:
![T1w_distribution](https://github.com/spinalcordtoolbox/disc-labeling-hourglass/assets/68945192/a72500e7-0bff-44be-a0fb-daee4412524a)

For T2w images:
![T2w_distribution](https://github.com/spinalcordtoolbox/disc-labeling-hourglass/assets/68945192/672a4cf6-9785-4e8b-8e9c-877296c699d0)

**Comment by jcohenadad** [comment]
But the spine-generic dataset should not be used as an absolute representation of scans around the world. Some hospitals acquire only 4-5 vertebrae, others acquire 10-15. There is no general rule about the number of discs to expect in one MRI scan unfortunately.

**Comment by NathanMolinier** [comment]
Yes I agree, but I still need to choose a relevant number of class for the training to limit false detections caused by under-represented discs.

Moreover, I might need to rewrite some functions to handle images with a different FOV.

**Comment by NathanMolinier** [comment]
The hourglass is performing poorly on new datasets when trained only on the spinegeneric dataset. Indeed, for most of the discs, several predictions are created. Further investigation need to be done regarding input parameters (resolution, FOV...) to detect potential differences in the input data.

**Comment by NathanMolinier** [comment]
I just noticed that the loss was only computed on non-empty GT masks resulting to a lot of false positive prediction.

https://github.com/spinalcordtoolbox/disc-labeling-hourglass/blob/1c8ff2893f85883ac94c21ec72257f0af4194009/src/dlh/models/jointsmseloss.py#L31-L37

This feature should be removed, however, for now I will just set the variable `use_target_weight` to False.

https://github.com/spinalcordtoolbox/disc-labeling-hourglass/blob/1c8ff2893f85883ac94c21ec72257f0af4194009/src/dlh/models/jointsmseloss.py#L15-L19
